libs/ledger-live-common/src/deviceSDK/tasks/updateFirmware.ts

```diff
@@ -0,0 +1,283 @@
+import { LockedDeviceError } from "@ledgerhq/errors";
+import { log } from "@ledgerhq/logs";
+import type {
+  DeviceId,
+  DeviceInfo,
+  FirmwareInfo,
+  FirmwareUpdateContext,
+} from "@ledgerhq/types-live";
+
+import { getVersion } from "../commands/getVersion";
+
+import ManagerAPI from "../../api/Manager";
+import { withDevice, withDevicePolling } from "../../hw/deviceAccess";
+import { from, Observable, of, Subscriber } from "rxjs";
+import { map, switchMap } from "rxjs/operators";
+import { SharedTaskEvent, sharedLogicTaskWrapper } from "./core";
+import { installFirmwareCommand } from "../commands/firmwareUpdate/installFirmware";
+import Transport from "@ledgerhq/hw-transport";
+import getDeviceInfo from "../../hw/getDeviceInfo";
+import { flashMcuOrBootloaderCommand } from "../commands/firmwareUpdate/flashMcuOrBootloader";
+import { quitApp } from "../commands/quitApp";
+
+export type UpdateFirmwareTaskArgs = {
+  deviceId: DeviceId;
+  updateContext: FirmwareUpdateContext;
+  // TODO: check if we should receive this here or rather retrieve it from the api in the task
+};
+
+export type UpdateFirmwareTaskError =
+  | "DeviceOnDashboardExpected"
+  | "DeviceOnBootloaderExpected"
+  | "McuVersionNotFound"
+  | "TooManyMcuOrBootloaderFlashes";
+
+export type UpdateFirmwareTaskEvent =
+  | { type: "installingOsu"; progress: number }
+  | { type: "flashingMcu"; progress: number }
+  | { type: "flashingBootloader"; progress: number }
+  | { type: "firmwareUpdateCompleted" }
+  | { type: "installOsuDevicePermissionRequested" }
+  | { type: "installOsuDevicePermissionGranted" }
+  | { type: "allowSecureChannelRequested" }
+  | { type: "taskError"; error: UpdateFirmwareTaskError }
+  | SharedTaskEvent;
+
+function internalUpdateFirmwareTask({
+  deviceId,
+  updateContext,
+}: UpdateFirmwareTaskArgs): Observable<UpdateFirmwareTaskEvent> {
+  return new Observable((subscriber) => {
+    withDevice(deviceId)((transport) =>
+      quitApp(transport).pipe(
+        switchMap(() => getVersion(transport)),
+        switchMap((value) => {
+          if (value.type === "unresponsive") {
+            return of({
+              type: "error" as const,
+              error: new LockedDeviceError(),
+            });
+          }
+          const { firmwareInfo } = value;
+
+          // TODO: we're repeating the handling of the unresponsive event here... is there a way to make this better
+          // the problem that if we want to change multiple commands that may have unresponsive events
+          // and only continue with the next command if the event is not unresponsive we might have to handle multiple times
+          // the unresponsive event :/
```

Comment:
Yes that the problem of handling this kind of "locked device" (unresponsive from a command) at the command level. But as we discussed, we don't want this unresponsive strategy on every commands (for ex `getAppAndVersion` does not need it) and we want to customized the timeout.

Though maybe we can create a custom `switchMap` and `map` that already handle the `unresponsive` event ? WDYT ? Is it just creating a new abstraction for just 5 lines of code ?

Reply:
I'm not completely against it. But I usually prefer to create the abstraction the moment it's actually needed (e.g. when we've created several tasks already that do this).

Reply:
yeah i agree - then let's keep it like this (handling manually the event of type `unresponsive`) and see if this pattern is repeated everywhere